### ISSUE TITLE
Review - 2020.04.05

### ISSUE THREAD

**Issue by JinhangZhu** [JinhangZhu/project-diary#5]
![week3.png](https://i.loli.net/2020/04/06/cjZO5HoTvt2kfqd.png)
## Done

- [x] 学习莫烦的PyTorch教程基础

- [x] 上手YOLOv3检测过程的代码

### `3.30-4.2`
考虑到第一次用[官方的PyTorch入门文档](https://pytorch.org/tutorials/beginner/deep_learning_60min_blitz.html)根本就没怎么懂神经网络的构建方法，所以重新按照[莫烦Python的教程](https://morvanzhou.github.io/tutorials/machine-learning/torch/)，从脚踏实地的**coding**直接开始。
- [PyTorch基础](https://github.com/JinhangZhu/pytorch-60min/tree/master/morvanpython/2_pytorch_nn_basics)，包括1) tensor和numpy转换，2) tensor的Variable（但是现版本不需要使用Variable API，见[Variable (deprecated)]()https://pytorch.org/docs/stable/autograd.html#variable-deprecated），3) 主要的激活函数。

- [Regression & Classification和一般的神经网络搭建](https://github.com/JinhangZhu/pytorch-60min/tree/master/morvanpython/3_pytorch_first_nn)。这一节挺重要，直接把我拉入神经网络coding的大门，值得反复回看。
- - 重点内容是：[**Mini-batch gradient descent**](https://github.com/JinhangZhu/pytorch-60min/blob/master/morvanpython/3_pytorch_first_nn/305_batch_train.py)，其中主要看**如何制作dataset**的代码段。
- - 重点复习思维导图和网络搭建的代码。
- - 其次多复习matplotlib做图。
其他内容有： 1)保存和读取，2)优化器。

- [MNIST手写数字的卷积神经网络](https://github.com/JinhangZhu/pytorch-60min/tree/master/morvanpython/4_advanced_nn)，重点为：
- - 训练的循环段的代码。
- - CNN网络搭建的代码。
- - 训练集和测试集制作的代码。
好像全都包括进来了呢😂，不过确实都需要背下来，熟练掌握。

### `4.3`
- 正式启动系统地对CNN的学习，教材采用李菲菲的课程材料Module 2[Convolutional Neural Networks: Architectures, Convolution / Pooling Layers](https://cs231n.github.io/convolutional-networks/)，前面的Modules是基础，后面也要反复查阅[CS231n Convolutional Neural Networks for Visual Recognition](https://cs231n.github.io/)。这一天把卷积层，池化层（即下采样层）仔细学习了，要多复习[CNN的思维导图](https://www.xmind.net/embed/a3Rfh3/)。

### `4.4`
- 开始写YOLOv3的代码。当日实现了两主体：
- - 通过读取官方`yolov3.cfg`文件，建立一个`module_list`，包括每一层的`module`实体，共106层，与论文👇一致。
![](https://i.loli.net/2020/04/04/T4yxQ3btrvhISog.png)
- - 实现了网络的architecture（不含yolo检测层），即名为`Darknet`的类定义，包括`__init__()`和`forward()`两个函数。其实整个网络能直接给输入按照`torch`给定的功能输出的就convolutional卷积层和上upsample采样层，其余层（route layer, shortcut layer, yolo/detection layer）都是需要自己写的。

### `4.5`
- 实现了：给定一张图片（416 * 416），能够执行一侧forward pass，输出predictions，形式是一个3D的tensor，内层的2D中的每一行是一个bbox的85个属性（如果对于COCO集），对于该图像输出即`torch.Size([1, 10647, 85])`。
```
    BBox per cell   ->  index in (batch, number of cells * number of anchors per cell, number of attributes)
    1st BBox        ->  (0, 000, 5+num_classes)
    2nd BBox        ->  (0, 001, 5+num_classes)
    3rd BBox        ->  (0, 002, 5+num_classes)
    1st BBox        ->  (0, 010, 5+num_classes)
    2nd BBox        ->  (0, 011, 5+num_classes)
    3rd BBox        ->  (0, 012, 5+num_classes)
    ...
    1st BBox        ->  (bs-1, [gs-1][gs-1]0, 5+num_classes)
    2nd BBox        ->  (bs-1, [gs-1][gs-1]1, 5+num_classes)
    3rd BBox        ->  (bs-1, [gs-1][gs-1]2, 5+num_classes)
```

## Next
- 继续完成YOLOv3代码的coding，week4之内要完成。
- 继续学习[ConvNet Architectures](https://cs231n.github.io/convolutional-networks/#architectures)
- 本周的项目进展到此为止，明天就要把上周的 #4 对YOLOv2的两个疑问回答一下。

**Comment by JinhangZhu** [comment]
How to get anchors from a dataset?

- [目标检测算法之YOLO系列算法的Anchor聚类代码实战](https://zhuanlan.zhihu.com/p/95291364)
